### ISSUE TITLE
allow filter parameter to pass an array with globbing

### ISSUE THREAD

**Issue by oupala** [ksmithut/metalsmith-uglify#11]
The `concat` option does not allow to pass an array. In my case, I would like to be able to have the following configuration :

``` 
filter: ['js/jquery-2.1.1.min.js', 'js/bootstrap/*.js', 'js/**/*.js']
```

and I get the following error :

``` 
Error: uglify: Failed to minify : TypeError: Cannot read property 'figure_out_scope' of null
```

**Comment by oupala** [comment]
metalsmith-concat had the same type of issue : aymericbeaumet/metalsmith-concat#9

**Comment by oupala** [comment]
In fact, passing an array does not seem to work at all. The two following statements do not work and output the same error message as above :

```
filter: ['js/test/myfile.js'],
filter: ['js/test/*.js'],
```

**Comment by ksmithut** [comment]
I'll take a look at this this weekend. Sorry it couldn't be sooner... been pretty hectic at my day job :) Should be a pretty straightforward fix, though

**Comment by oupala** [comment]
My latest tests show that the *array of string* do not work at all (nothing is minified) but might not output an error (I changed node version from 0.10x to 0.12.x, which might explain the new behaviour).

Anyway, the important thing is that nothing is minified.

**Comment by ksmithut** [comment]
that's good to know. Apologies for not getting to this. Life has gotten hectic. If you're up to it, a PR would be greatly appreciated, otherwise I might not be able to get around to fixing it until Saturday or sometime next week.

**Comment by oupala** [comment]
I would be happy to create a pull request for this issue. Unfortunately I do not have the technical level to be able to do this.

I hope to be able one day, but this is not yet the case...

I'll have to wait for your fix, whatever its publication date.

**Comment by ksmithut** [comment]
No worries, Thanks for your patience

**Comment by ksmithut** [comment]
Alright, I'm working on this now

**Comment by ksmithut** [comment]
@oupala I've published a new version (1.2.0). could you give it a try when you get a chance?

**Comment by oupala** [comment]
This release works very very well.

I can finally enable uglification on my website.

Thanks @ksmithut !

**Comment by oupala** [comment]
I am maybe misleading with the use of `metalsmith-uglify`.

The standard usecase is maybe to concatenate with `metalsmith-concat`, then uglify with `metalsmith-uglify`.

@ksmithut can you confirm if `metalsmith-uglify` should be used (or not) for concatenating (in addition to uglifying)?

**Comment by ksmithut** [comment]
@oupala You should be able to use it in this way. Could you give an example of your folder structure and metalsmith middleware usage? If you could give me the least amount of code and setup it takes to reproduce the error, that would help me be able to write a test case to solve.

**Comment by oupala** [comment]
I'll try to explain you the minimal configuration that I need.

To be brief, the main problem is how to set the order of file concatenation.

This is an extract from my build configuration:

```js
.use(uglify({
          filter: ['js/b.js', 'js/*.js'],
          concat: 'js/app.min.js'
        }))
```

And this is my folder structure:

```
.
├── js
│   ├── a.js
│   ├── b.js
│   ├── c.js
```

When **metalsmith-uglify** is concatenating my js file, I said that I wanted that `b.js` has to be concatenated **first**, then all other files should be concatenated also.

The algorithm should looks like that:

1. concatenate `js/b.js` to app.min.js
2. list all files that match `js/*.js`
3. remove `js/b.js` from the list as it has already been concatenated at step 1
4. concatenate all files from the list to app.min.js
    - this means concatenate `js/a.js` and `js/c.js` to app.min.js
    - *or*
    - this means concatenate `js/c.js` and `js/a.js` to app.min.js
    - as there was nos preferences in the order of files when using `js/*.js`, there is no importance of the order of concatenating `js/a.js` and `js/c.js`

**Comment by oupala** [comment]
My problem is that I need to concatenate JQuery, Boostrap, and my personal js files.

As I suppose that your plugin is concatenating the files in the alpahbetical order, Bootstrap appears before JQuery in the app.min.js. The problem is that Bootstrap needs JQuery in order to be instanciated correctly.

That why I need to specify an order between files concatenation.

**Comment by oupala** [comment]
The example above is simple an minimalistic.

But the algorithm should work for more complicated usecases such as the example below:

```js
        .use(uglify({
          filter: ['js/jquery-2.1.1.min.js', 'js/bootstrap/*', 'js/**/*.js'],
          concat: 'js/app.min.js'
        }))
```

In this example, I say that I want to following order for concatenation:

1. conatenate jquery
2. concatenate all files that are in the bootstrap directory
3. concatenate all files that are in the js directory
    - but **not** the jquery file which has already been concatenated
    - but **not** the files that are in the bootsrtap directory which have already been concatenated

I hope this is clear enough for you to understand.

**Comment by ksmithut** [comment]
@oupala Ah I understand now what the issue is. Yes, this module does not order this in the order of the filtering. I'll add that in. I do think that should probably be the job of `metalsmith-concat`, though, but it should be easy enough to add that in.

**Comment by oupala** [comment]
@ksmithut It's great if you can add this feature.

I tried to replace `metalsmith-uglify` by the combination of `metalsmith-concat` **and** `metalsmith-uglify`. First thing was to improve the `metalsmith-concat` plugin by resolving issues aymericbeaumet/metalsmith-concat#1 and aymericbeaumet/metalsmith-concat#9.

I was then able to use `metalsmith-concat` to concatenate all my files, and `metalsmith-uglify` to minify them. But I faced a new problem as I'm using sourcemaps. The sourcemaps generated by `metalsmith-uglify` were referring to my concatenated file instead of my real source files.

That's were I went back to `metalsmith-uglify` and created the current issue.

Maybe, it would have been a better idea to combine `metalsmith-uglify` **then** `metalsmith-concat`. What do you think of that?

**Comment by ksmithut** [comment]
That could definitely work. You might miss out on a few little
optimizations, but not enough to really make a difference. Gzipping your
assets is going to save you the most space. I've got those changes ordering
the files, but it's a separate `order` option to separate filtering and
ordering concerns. I'm working on another issue as we speak and once I
finish that one, I'll push out another release.

**Comment by oupala** [comment]
Should I also concat sourcemaps?

As you can see, combining `metalsmith-uglify` then `metalsmith-concat` might more difficult than it appears at first sight...

**Comment by oupala** [comment]
Are you still planning to fix this issue? I just want to know if I can wait for it in a near future.

Thanks for your reply.

**Comment by ksmithut** [comment]
Yes, Apologies, been swamped the past few weeks. I'll work on it this
morning and hopefully have something ready this week.

**Comment by ksmithut** [comment]
@oupala Are you using metalsmith-uglify to generate a single javascript file? I'm trying to figure out if I should be doing concatenating, or if I should somehow take in sourcemap input options. So you would run metalsmith-concat first, it would generate source maps, and then you would pass in some other option to metalsmith-uglify to use those sourcemaps on input. Unfortunately, it doesn't look like metalsmith-concat generates sourcemaps...

It's looking more and more like metalsmith-uglify will need to support concat with sourcemaps for fully. I'll look into this today and tomorrow.

**Comment by oupala** [comment]
As far as I know, metalsmith-concat does not generate sourcemaps. And I don't see why it should, as sourcemap is only useful when uglifying. metalsmith-concat could be used to concat anything, not only js and css code.

This is why I think metalsmith-uglify should support **concat** *and* **sourcemap**.

In my usecase, I would pass some js files to metalsmith-uglifying **in a specific order**, and get a unique output file, with its associated sourcemap.

And last but not least, I would also fingerprint my js file, so that it is well managed by cache servers. But I think, metalsmith-fingerprint does not support sourcemap (which should also be renamed).

In my opinion, metalsmith-uglify should also support fingerprinting.

This orientation, is not really compliant with metalsmith principles where eash plugin does one thing, but well. How can web managed this when tasks are tightly linked such as uglifying, concatenating, sourcemapping, fingerprinting ?

It is possible to make metalsmith-uglify call subplugins such as metalsmith-concat, metalsmith-sourcemap and metalsmith-fingerprint ?

**Comment by ksmithut** [comment]
In regards to sourcemap support, metalsmith-concat should support it. If it takes input files and generates a different file, it should support sourcemaps (although it doesn't appear they do as you mentioned). Grunt/Gulp plugins support it, so metalsmith plugins should as well.

Fingerprinting should definitely support sourcemaps, even if they don't. Most plugins that do support sourcemaps have some kind of input sourcemap option, which takes the input source map and generates a new sourcemap, which is what I'm trying to support, but having difficulty with.

For front end build process management, I might try something like https://github.com/jelz/gulp-metalsmith to have much better sourcemap support, as well as concatting, fingerprinting, and sourcemaps. I'll continue to work on this piece of the metalsmith pipeline, but with the lack of sourcemap support with other plugins, metalsmith may be an incomplete solution to the web asset build process problem.

**Comment by oupala** [comment]
For sourcemap support in metalsmith-concat, I do mention @aymericbeaumet (I filed an issue about that: aymericbeaumet/metalsmith-concat#15).

For sourcemap support in metalsmith-fingerprint, I do mention @@christophercliff (I also filed an issue about that: christophercliff/metalsmith-fingerprint#10).

**Comment by oupala** [comment]
@ksmithut do you still plan to add the **concat by order** feature to your plugin?

**Comment by ksmithut** [comment]
Yes, Concat by order is supported in the latest version, just not with sourcemaps. Uglify's sourcemap support is not as good as I'd hoped, and even less so when trying to combine concat and uglify sourcemapping.

For the ordering, use the `options.order` option described in the README.md. I still plan on being able to support sourcemaps, but I'm in crunch time at work.

**Comment by oupala** [comment]
I don't understand how to use `options.order`. The file `metalsmith-uglify-test.js` does not give me enough clues.

I want to concatenate jQuery before other js files. How should I do that ?

```js
        .use(uglify({ // minify and concat javascript
          //filter: ['js/jquery-2.1.1.min.js', 'js/**/*.js'],
          filter: ['js/**/*.js'],
          preserveComments: 'some',
          removeOriginal: true,
          sourceMap: true,
          concat: 'js/app.min.js',
          order: ['js/jquery-2.1.1.min.js', 'js/**/*.js']
        }))
```

**Comment by ksmithut** [comment]
yup, that should do it. Is it not working for you with those options?

**Comment by oupala** [comment]
Nope, unfortunately, it doesn't.

Here is my js directory:

```
js
¦-- bootstrap.js
¦-- highlight.pack.js
¦-- jquery-2.1.1.js
¦
+---custom
    ¦-- application.js
    ¦-- browser-update.js
    ¦-- gotop.js
    ¦-- jquery.tableofcontents.js
    ¦-- tocgenerator.js
```

I've updated metalsmith-uglify to version 1.2.1.

And bootstrap is still in first position in the concatenated file.

**Comment by oupala** [comment]
If I only leave the following file in my `js` directory (I just remove the `custom` directory):

```
js
¦-- bootstrap.js
¦-- highlight.pack.js
¦-- jquery-2.1.1.js
```

The concatenated file is the following result:

```
bootstrap,highlight,jquery;
//# sourceMappingURL=app.min.js.map
```

**Comment by oupala** [comment]
I tried many other configurations, such as:

```js
        .use(uglify({ // minify and concat javascript
          filter: ['js/**/*.js'],
          preserveComments: 'some',
          removeOriginal: true,
          sourceMap: true,
          concat: 'js/app.min.js',
          order: ['js/jquery-2.1.1.min.js']
        }))
```

But it still doesn't work. It looks like the `sort` has no effect.

I also checked the test file, and I don't understand how the test file achieve testing *ordering*, wile I think it just test the **presence** of the required file, not its **order**.

**Comment by ksmithut** [comment]
I'll keep trying it out... If you've got a repo or gist that I can look at your setup to replicate your issue, that would help out a ton. Don't publish anything too revealing, just the minimum amount of code and setup it takes to replicate the issue.

**Comment by oupala** [comment]
Hi @ksmithut,

I've just created a testing repo for *concat by order* usecase.

See this [test repo](https://github.com/oupala/uglify-test).

I hope this will help you!

**Comment by julienvanderkluft** [comment]
Hi @ksmithut and @oupala,

I have to work on project with Metalsmith these days, and obviously i use `metalsmith-uglify` for my JS files. I was little worried when i read this issue. But finally, everything work fine. Take a look on my settings @oupala if it's can help you : 

```
.use(uglify({
        filter: ['js/**/*.js'],
        removeOriginal: true,
        sourceMap: true,
        order: ['js/vendors/**/*.js', 'js/*.js'],
        concat: 'js/main.min.js'
      }))
```

For a JS directory like this :
```
js
+---vendors
    ¦-- jquery.js
¦-- app.js
```

**Comment by oupala** [comment]
Thanks @julienvanderkluft for this working configuration. And sorry @ksmithut for my bad testing.

I am ashamed because my test case was wrong...

My test case was trying to give the priority to the file `js/jquery-2.1.1.min.js` while my source directory was containing the file `jquery-2.1.1.js`. Did you see the `.min` difference between the two filename? I did not until looking at it again after @julienvanderkluft message.

I'm sooo sorry!

**Comment by ksmithut** [comment]
No worries, I apologize for not being super responsive. It's been crazy at work... I'll be able to be more responsive to future issues now that work craziness is almost done.
